### ISSUE TITLE
Device motion API exposure

### ISSUE THREAD

**Issue by subrina01** [bromite/bromite#160]
**Is your feature request related to privacy?**
Yes

**Is there a patch available for this feature somewhere?**
Not that I'm aware of.

**Describe the solution you'd like**
Block or fake all events coming from:
window.DeviceMotionEvent
window.Accelerometer
window.Gyroscope
window.Magnetometer
window.LinearAccelerationSensor
window.GravitySensor

You can see the data leaking on this web site:
https://whatwebcando.today/device-motion.html

**Describe alternatives you've considered**
Tried to find a Chrome flag to disable all sensors above, didn't find such option in chrome-flags.

**Comment by subrina01** [comment]
Link to relevant source code:

Sensor API:
https://github.com/chromium/chromium/tree/master/third_party/blink/renderer/modules/sensor

Device orientation API:
https://github.com/chromium/chromium/tree/master/third_party/blink/renderer/modules/device_orientation

Generic sensor API:
https://github.com/chromium/chromium/tree/master/services/device/generic_sensor

**Comment by DecodeMehh** [comment]
window.DeviceOrientationEvent
window.AbsoluteOrientationSensor
window.RelativeOrientationSensor

**Comment by csagan5** [comment]
@antonin-lebrard I am using `70.0.3538.71` and I do see the device motion API values being returned. I wonder why in your case it's all blanked out? Are you using some special Android setting?

I noticed the existance of this API some time ago, however I am surprised that it does not ask for permission like in the case of camera/microphone.

> Also memory
> navigator.deviceMemory

@DecodeMehh you can open a separate issue to discuss that, this is about the device motion APIs.

@Eloston what is your take on this? What would be the privacy concerns and what is the best strategy to follow? I can probably have a `chrome://flags` flag to disable them globally, but it's going to be hardly good in case of HTML5 games (the only valid use of this API I can think of).

**Comment by antonin-lebrard** [comment]
Oh I thought this was global, sorry for the inconvenience.
I touched the flags of bromite at some point if I recall correctly, but I have abandonned to touch them since I though they were overriden at each updates. Maybe they have been kept since.

And indeed, after seeing the list of flags on chrome://version Command line part, I have this list of `--disable-features`:
- GamepadExtensions
- GenericSensor 
- GenericSensorExtraClasses
- ImageCaptureAPI
- OfflinePagesPrefetching
- OmniboxSpeculativeServiceWorkerStartOnQueryInput
- UserMediaScreenCapturing
- WebXR
- WebXROrientationSensorDevice

The GenericSensor, GenericSensorExtraClasses, WebXR, WebXROrientationSensorDevice seems to be related.

**Comment by subrina01** [comment]
Some additional observations:

I have disabled GenericSensor, GenericSensorExtraClasses, WebXR, WebXROrientationSensorDevice in chrome flags, but window.DeviceMotionEvent events are still leaking correct values, as seen on demo website above. They must be controlled by another flag, or no flag at all (and lacking permissions control).

@Eloston @csagan5 The privacy concerns are investigated in several papers:

"Tracking Mobile Web Users Through Motion Sensors: Attacks and Defenses"
https://pdfs.semanticscholar.org/aeb3/fc01c0c834b9d64c3cf5c3a1e0e499326dbb.pdf

"Smartphone Fingerprinting Via Motion Sensors: Analyzing Feasibility at Large-Scale and Studying Real Usage Patterns"
https://arxiv.org/pdf/1605.08763.pdf

From the second paper:
"CONCLUSION
We demonstrated that sensor fingerprinting is feasible on a much
larger scale than previously studied. We showed that 90% accuracy
can be achieved for up to 400 devices, and at least 12–16% accuracy
can be realized with 100 000 devices, as predicted according
to our model. "

**Comment by Eloston** [comment]
@csagan5 If I were to make a decision here, I'd need to first look for the consequences of leaving these sensor values exposed. **In general, how can these APIs be used to invade privacy?** For example, are there any specific algorithms that can use these APIs to fingerprint the user? Could these APIs be used to determine something about the user that isn't explicitly given by the user (e.g. location, movement, etc.)? Could the data from these APIs be combined with other metrics to extrapolate something meaningful?

Based on my knowledge, I believe these sensors are too nosiy and specific to extrapolate anything useful about the user. Many wearable electronics include these sensors, but I have yet to see anyone deduce anything significant about the user just using the data from these sensors. I realize this is a pretty naive argument, so if anyone has found any significant results using these sensors' data (or a proof of why no such results can exists), I will be interested to hear it.

**Comment by Eloston** [comment]
@csagan5 @subrina01 I skimmed both papers and it seems that manipulating the sensor values is an effective countermeasure to fingerprinting for these sensors.

*"Smartphone Fingerprinting Via Motion Sensors: Analyzing Feasibility at Large-Scale and Studying Real Usage Patterns"*
 proposes quantization (not as effective) and noise (more effective) as countermeasures, while *"Tracking Mobile Web Users Through Motion Sensors: Attacks and Defenses"* has several different noise countermeasure methods.

A little off-topic, but the highlight of these two papers is using an inaudible tone from the speakers to stimulate a unique-enough response from the gyroscope sensor in *"Smartphone Fingerprinting Via Motion Sensors: Analyzing Feasibility at Large-Scale and Studying Real Usage Patterns"*. It sounds far-fetched but it's fascinating to see it work.

**Comment by Eloston** [comment]
@csagan5 

> I think with #enable-generic-sensor we have already defeated most of what is covered in the papers, I am interested in the remaining APIs (and why they are still enabled...if there is no flag to control them).

For the websites that do have access to the APIs, there's still a potential for fingerprinting to happen in the background. Implementing noise would help here. Granted, I don't really know if there are any websites that are like this. However, if the device motion API doesn't have a toggle, adding noise may be an easier solution.

**Comment by csagan5** [comment]
@Eloston the potential to fingerprint is there only if the flags are enabled.

My strategy here would be:
* first have a flag to disable (I am waiting to see if I am successful with adding the missing device motion flag) the API, and have it disabled by default
* then see if another option to add noise can be implemented; if this works, it could be reconsidered whether to leave it enabled or not.

I think other flags have this tri-state as well; I wonder how complex it would be to add noise e.g. what is the prevision level and significant digits, I would have to read the papers first.

@subrina01 I am still waiting for your reply on the points in [this comment](https://github.com/bromite/bromite/issues/160#issuecomment-431946136); if there are other related vectors or aspects not considered, they should be for a mitigation to this issue.

**Comment by csagan5** [comment]
Firefox has already taken steps for these APIs: https://bugzilla.mozilla.org/show_bug.cgi?id=1359076

(this was found from the related https://github.com/w3c/sensors/issues/182)

I have successfully created a patch to disable the motion API so it will be disabled by default in Bromite starting from next release.

In general in Bromite I try to follow:
1. privacy first
2. features second

It is because these two have been inverted that we are in the current situation anyways...

From https://arxiv.org/pdf/1605.08763.pdf (2016 paper):
> CONCLUSION
We demonstrated that sensor fingerprinting is feasible on a much
larger scale than previously studied. We showed that 90% accuracy
can be achieved for up to 400 devices, and at least 12–16% accuracy
can be realized with 100 000 devices, as predicted according
to our model. Our measurement study reveals that motion sensors
are already used by over 1% of the top 100 000 websites, and that
sensor data are often sent to servers, which could serve as a vehicle
for fingerprinting. Thus we can conclude that motion sensor
fingerprinting is a realistic threat to mobile users’ privacy.
We also evaluated the tradeoff between privacy and utility as
realized by two different fingerprinting mitigation strategies. Our
measurement study suggests that many applications of sensor data
are unlikely to be affected. Our user study shows that even for sensitive
applications that use motion sensors as control input, there
is no significant impact of privacy mitigation techniques on the usability
of motion sensors in this context, according to both subjective
and objective metrics.

I believe mitigations against precise timing measurements have been implemented by major browsers since then.

I would need to read all the other papers and the defensive strategies they used/propose...I will leave that part out for now. I think implementing "pond poisoning" via noise (another proposed approach is quantisation) is feasible but slightly different than what was done for the canvas and the client rects patch.
